### PR TITLE
HPCC4J-562 Odd unsigned decimals incorrect scale

### PR DESCRIPTION
**Pull request by jpmcmu**
- Corrected BinaryRecordReader unsigned decimal parse code
- Updated ECL test datasets to include odd precision decimal values

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [X] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] This change does not cause any existing JUnits to fail.
  - [ ] I have include JUnit coverage to test this change
  - [ ] I have performed system test and covered possible regressions and side effects.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] This change fixes the problem, not just the symptom

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana please review. @jchambers-ln @drealeed FYI

File: dfsclient/src/test/resources/generate-datasets.ecl [hunk 1]
```diff
@@ -33,7 +36,9 @@ ds := DATASET(totalrecs1, transform(rec,
                                    self.r8 := (REAL)(random() % unique_values);
                                    self.r4 := (REAL)(random() % unique_values);
                                    self.dec16 := (REAL)(random() % unique_values);
+                                   self.dec15 := (REAL)(random() % unique_values);
                                    self.udec16 := (REAL)(random() % unique_values);
+                                   self.udec15 := (REAL)(random() % unique_values);
```
Comment (rpastrana) [inline dfsclient/src/test/resources/generate-datasets.ecl:41]: Is there a corresponding test to verify these odd len decimals are correctly read?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 2]
```diff
@@ -830,7 +830,11 @@ private BigDecimal getUnsignedDecimal(int numDigits, int precision, int dataLen)
         BigDecimal ret = new BigDecimal(0);
 
         int idx = 0;
-        int curDigit = numDigits - 1;
+        int curDigit = numDigits;
+        if ((numDigits % 2) == 0)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:834]: It's not entirely obvious to me why this logic is needed, perhaps an inline comment?

**Review by rpastrana — COMMENTED** [review]
@jpmcmu a couple of comments.

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java (hunk 2 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:838]: Good point, added a comment and squashed

File: dfsclient/src/test/resources/generate-datasets.ecl (hunk 1 above)
Comment (jpmcmu) [reply dfsclient/src/test/resources/generate-datasets.ecl:41]: @rpastrana Yes, tests in DFSReadWriteTest read those generated datasets, write them back, and then compare the datasets against the generated datasets for correctness.

**Review by rpastrana — APPROVED** [review]
@jpmcmu

**Pull request merged**